README.md
Modified RPG Monocular Pose Estimator
=====================================

PLEASE READ THROUGH CHANGES.md TO FIND OUT WHAT HAS CHANGED IN THIS REPOSITORY.

Disclaimer and License
----------------------

The modified RPG Monocular Pose Estimator has been tested under the following setups:

* ROS-Indigo and Ubuntu 14.04

This is research code, expect that it changes often and any fitness for a particular purpose is disclaimed.

The source code is released under a GPL licence. Please contact the authors for a commercial license.

Package Summary
---------------

The RPG Monocular Pose Estimator uses infrared LEDs mounted on a target object and a camera with an infrared-pass filter to estimate the pose of an object.

The positions of the LEDs on the target object are provided by the user in a YAML configuration file. The LEDs are detected in the image, and a pose estimate for the target object is subsequently calculated.

### Publication

If you use this work in an academic context, please cite the following [ICRA 2014 publication](http://rpg.ifi.uzh.ch/docs/ICRA14_Faessler.pdf):

M. Faessler, E. Mueggler, K. Schwabe, D. Scaramuzza: 
**A Monocular Pose Estimation System based on Infrared LEDs.**
IEEE International Conference on Robotics and Automation (ICRA), Hong Kong, 2014.

    @inproceedings{Faessler2014ICRA,
      author = {Faessler, Matthias and Mueggler, Elias and Schwabe, Karl and Scaramuzza, Davide},
      title = {A Monocular Pose Estimation System based on Infrared {LED}s},
      booktitle = {IEEE International Conference on Robotics and Automation (ICRA)},
      year = {2014}
    }

Watch the [video](http://www.youtube.com/watch?v=8Ui3MoOxcPQ) demonstrating the RPG Monocular Pose Estimator:   
[![ RPG Monocular Pose Estimator Video](http://img.youtube.com/vi/8Ui3MoOxcPQ/0.jpg)](http://www.youtube.com/watch?v=8Ui3MoOxcPQ)


Installation
------------

### Installation of the package

#### Dependencies

The RPG Monocular Pose Estimator is built on the Robotic Operating System (ROS). In order to install the package, ROS has to be installed.

- In order to install the Robot Operating System (ROS), please follow the instructions provided in the [link](http://wiki.ros.org).
- Make sure you have properly set up a ROS catkin workspace as described [here](http://wiki.ros.org/ROS/Tutorials/InstallingandConfiguringROSEnvironment).

Additionally, the RPG Monocular Pose Estimator makes use of [OpenCV](http://opencv.org) for image processing and the [Eigen](http://eigen.tuxfamily.org) linear algebra library. These should come preinstalled with ROS, however, if the dependency is missing they can be installed from their respective websites:

- To install OpenCV, follow the installation instructions provided on the OpenCV [website](http://opencv.org).

- To install the Eigen linear algebra library, follow the installation instructions provided on the Eigen [website](http://eigen.tuxfamily.org).

#### Main Installation

In order to install the RPG Monocular Pose Estimator, clone the latest version from our *GitHub* repository into your catkin workspace and compile the package using ROS.

    cd catkin_workspace/src
    git clone https://github.com/uzh-rpg/rpg_monocular_pose_estimator.git
    cd ../
    catkin_make

### Building the Documentation

The RPG Monocular Pose Estimator makes use of [Doxygen](http://www.doxygen.org/‎) to produce its documentation. Please ensure that you have the latest version of Doxygen (version 1.8.5). The installation instructions for the lastest version of Doxygen can be found [here](http://www.stack.nl/~dimitri/doxygen/download.html).

The Doxygen configuration file is located in the *doxygen_documentation* folder within the RPG Monocular Pose Estimator directory. 

Change to the doxygen_documentation directory:

    roscd monocular_pose_estimator/../doxygen_documentation

To produce the documentation, run Doxygen on the configuration file:

    doxygen doxygen_config_file.cfg

Open the index.html file in *html* directory that has been produced.

Test Installation on Basic Dataset
----------------------------------

In order to test the installation on a data set, download the data set from [here](http://rpg.ifi.uzh.ch/data/monocular-pose-estimator-data.tar.gz), and follow these instructions.

1.    Download and Untar a sample ROS bag file

          roscd monocular_pose_estimator
          mkdir bags
          cd bags
          wget http://rpg.ifi.uzh.ch/data/monocular-pose-estimator-data.tar.gz
          tar -zxvf monocular-pose-estimator-data.tar.gz
          rm monocular-pose-estimator-data.tar.gz

2.    Launch the demo launch file using

          roslaunch monocular_pose_estimator demo.launch

3.    You should see a visualisation of the system: detected LEDs should be circled in red, the region of interest in the image that is being processed should be bounded by a blue rectangle, and the orientation of the tracked object will be represented by the red-green-blue trivector located at the origin of the traget object's coordinate frame.


Basic Usage
-----------

The program makes use of a number of parameters to estimate the pose of the tracked object. These include the location of the LEDs on the object, the intrinsic parameters of the camera, and various runtime prameters, such as thresholding values.

#### Setting up the Marker Positions Parameter File

In order to predict the pose of the target object, the RPG Monocular Pose Estimator needs to know the positions of the LEDs on the object in the object's frame of reference. These positions are given to the program using a YAML file located within the `roscd monocular_pose_estimator/marker_positions` folder.

By default, the file is called `marker_positions.YAML` and has the following format.

    # The marker positions in the trackable's frame of reference:
    #
    marker_positions:
      - x: 0.0714197
        y: 0.0800214
        z: 0.0622611
      - x: 0.0400755
        y: -0.0912328
        z: 0.0317064
      - x: -0.0647293
        y: -0.0879977
        z: 0.0830852
      - x: -0.0558663
        y: -0.0165446
        z: 0.053473

**Note** that each new LED position is indicated by a dash (-). The position is given in the x, y, and z coordinates of the object frame in metres.

If you would like to use your own marker positions file, place it in the `monocular_pose_tracker/marker_positions` folder and alter the launch file as explained below in the section 'Launch Files'.


#### Running the RPG Monocular Pose Estimator with a USB Camera

Ensure that you have a working USB camera.

The camera needs to be calibrated. Follow the instructions at http://www.ros.org/wiki/camera_calibration/Tutorials/MonocularCalibration.

The pose estimator listens to the camera/image_raw topic and the  camera/camera_info topic of the camera at launch. An example of such a camera_info topic is:
    
    header: 
      seq: 4686
      stamp: 
        secs: 1378817190
        nsecs: 598124104
      frame_id: /camera
    height: 480
    width: 752
    distortion_model: plumb_bob
    D: [-0.358561237166698, 0.149312912580924, 0.000484551782515636, -0.000200189442379448, 0.0]
    K: [615.652408400557, 0.0, 362.655454167686, 0.0, 616.760184718123, 256.67210750994, 0.0, 0.0, 1.0]
    R: [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0]
    P: [525.978149414062, 0.0, 357.619310580667, 0.0, 0.0, 579.796447753906, 258.377118195804, 0.0, 0.0, 0.0, 1.0, 0.0]
    binning_x: 0
    binning_y: 0
    roi: 
      x_offset: 0
      y_offset: 0
      height: 0
      width: 0
      do_rectify: False
    ---

The camera should be adjusted so that the gain and shutter/exposure time of the camera are fixed. (Use 'rqt_reconfigure'  to adjust the camera parameters while the camera is running). Ideally, the LEDs should appear very bright in the image so that they can easily be segmented from the surrounding scene by simple thresholding. See 'Parameter Settings' below on how to change the thresholding value.

#### Inputs, Outputs

##### Configuration Files

The RPG Monocular Pose Estimator requires the LED positions on the target object. These are entered in a YAML file and are loaded at runtime. See 'Setting up the Marker Positions Parameter File' above for more details.

##### Subscribed Topics

The RPG Monocular Pose Estimator subscribes to the following topics:

* camera/image_raw ([sensor_msgs/Image](http://docs.ros.org/api/sensor_msgs/html/msg/Image.html))

  The image form the camera. The LEDs will be detected in this image. 

* camera/camera_info ([sensor_msgs/CameraInfo](http://docs.ros.org/api/sensor_msgs/html/msg/CameraInfo.html))

  The camera calibration parameters.

##### Published Topics

The RPG Monocular Pose Estimator publishes the following topics:
    
* estimated_pose ([geometry_msgs/PoseWithCovarianceStamped](http://docs.ros.org/api/geometry_msgs/html/msg/PoseWithCovarianceStamped.html))

     The estimated pose of the target object with respect to the camera.

* image_with_detections ([sensor_msgs/Image](http://docs.ros.org/api/sensor_msgs/html/msg/Image.html))

  The image with the detected LEDs cirlced in red, the region of interest of the image that was processed bounded by a blue rectangle, and the orientation trivector projected onto the object.

#### Parameter Settings

The following parameters can be set dynamically during runtime. (Use 'rqt_reconfigure' to adjust the parameters).

exit(gen.generate(PACKAGE, "monocular_pose_estimator", "monocular_pose_estimator_params"))


* threshold_value (int, default: 220, min: 0, max: 255)

  This is the pixel intensity that will be used to threshold the image. All pixels with intensity values below this value will be set to zero. Pixels with intensity values equal to or higher than this value will retain their intensities.

* blur_size (double, default: 2, min: 0, max: 6)

  This is the size of the applied box blur in pixels.

* min_blob_area (double, default: 10, min: 0, max: 100)

  This is the minimum blob area (in pixels squared) that will be detected as a blob/LED.

* max_blob_area (double, default: 200, min: 0, max: 1000)

  This is the maximum blob area (in pixels squared) that will be detected as a blob/LED. Blobs having an area larger than this will not be detected as LEDs.

* max_width_height_distortion (double, default: 0.5, min: 0, max: 1)

  (REMOVED)
  This is a parameter related to the circular distortion of the detected blobs. It is the maximum allowable distortion of a bounding box around the detected blob calculated as the ratio of the width to the height of the bounding rectangle. Ideally the ratio of the width to the height of the bounding rectangle should be 1. 

* max_circular_distortion (double, default: 0.5, min: 0, max: 1)
  
  (REMOVED)
  This is a parameter related to the circular distortion of the detected blobs. It is the maximum allowable distortion of a bounding box around the detected blob, calculated as the area of the blob divided by pi times half the height or half the width of the bounding rectangle. 

* back_projection_pixel_tolerance (double, default: 3, min: 0, max: 10)

  This is the tolerance (in pixels) between a back projected LED and an image detection. If a back projected LED is within this threshold, then the pose for that back projection is deemed to be a possible candidate for the pose.

* nearest_neighbour_pixel_tolerance (double, default: 5, min: 0, max: 10)

  This is the tolerance for the prediction of the correspondences between object LEDs and image detections. If the predicted position of an LED in the image is within this tolerance of the image detections, then the LED and image detection are considered to correspond to each other.

* certainty_threshold (double, default: 0.75, min: 0, max: 1)

  This is the proportion of back projected LEDs into the image that have to be within the back_projection_pixel_tolerance, for a pose to be considered valid.

* valid_correspondence_threshold (double, default: 0.7, min: 0, max: 1)

  This is the ratio of all combinations of 3 of the correspondences that yielded valid poses (i.e., were within the certainty_threshold), for a set of correspondences to be considered valid.

* roi_border_thickness (int, default: 10, min: 0, max: 200)

  This is the thickness of the border (in pixels) around the predicted area of the LEDs in the image that defines the region of interest for image processing and detection of the LEDs.

#### Launch Files

The RPG Monocular Pose Estimator needs to be launched with a launch file, since the location of the YAML configuration file containing the LED positions on the object needs to be specified. (See 'Setting up the Marker Positions Parameter File' above for further details). An example launch file is presented below.

    <launch> 
        
    	<!-- Name of the YAML file containing the marker positions -->
        <arg name="YAML_file_name" default="marker_positions"/>

    	<!-- File containing the the marker positions in the trackable's frame of reference -->
		<arg name="marker_positions_file" default="$(find monocular_pose_estimator)/marker_positions/$(arg YAML_file_name).yaml"/> 

	    <node name="monocular_pose_estimator" pkg="monocular_pose_estimator" type="monocular_pose_estimator" respawn="false" output="screen"> 
		    <rosparam command="load" file="$(arg marker_positions_file)"/>
    		<param name= "threshold_value" value = "140" />
    		<param name= "gaussian_sigma" value = "0.6" />
    		<param name= "min_blob_area" value = "10" />
    		<param name= "max_blob_area" value = "200" />
    		<param name= "max_width_height_distortion" value = "0.5" />
    		<param name= "max_circular_distortion" value = "0.5" />
    	</node>
    </launch>

The name of the YAML configuration file may be specified directly as the default value of the argument `YAML_file_name`, or may be passed to the system via the command line on launch, e.g. `roslaunch monocular_pose_estimator launch_file_name.launch YAML_file_name:=<file_name>`, where `<file_name>` is replaced with the name of the maker positions YAML file.

This example launch file also sets some of the parameters described in 'Parameter Settings' above. They needn't be set in the launch file as, they can be dynamically changed during runtime. (Use 'rqt_reconfigure' to adjust the parameters).

For more information on how to use ROS launch files, see the ROS [website](http://wiki.ros.org/ROS/Tutorials/Roslaunch%20tips%20for%20larger%20projects).

Hardware
--------

The RPG Monocular Pose Estimator makes use of infrared LEDs mounted on the target object and a camera fitted with an infrared-pass filter. The details of the LEDs and camera that were used in our evaluation of the package are outlined below. Our system will be able to work with any kind of LEDs and camera, provided that the LEDs appear bright in the image and can consequently be segmented from the image using simple thresholding operations. Also the emission pattern of the LEDs should be as wide as possible, so that they still appear bright in the image even when viewed from shallow viewing angles.

### Infrared LEDs

The infrared LEDs that were used were SMD LEDs from Harvatek HT-260IRPJ of type Harvatek HT-260IRPJ. They have a wide emission pattern and consequently can be viewed from varying view points.

#### LED Configuration

The placement of the LEDs on the target object can be arbitrary, but must be non-symmetric. The LEDs should also not lie in a plane in order to reduce the ambiguities of the pose estimation. Also the larger the volume that the LEDs fill, the more accurate the estimated pose will be. The robustness and accuracy are improved if the LEDs are visible from many view points. The RPG Monocular Pose Estimator requires a minimum of four (4) LEDs mounted and visible on the target object in order for its pose to be resolved. 

### Camera

The camera used was a [MatrixVision](http://www.matrix-vision.com/) mvBlueFOX-MLC200w monochrome camera fitted with an infrared-pass filter. Its resolution is 752x480 pixels.


CHANGES.md
# Changes from the original code

This document lists the changes that have been committed in rapyutas repository
in order to speed up the monocular pose estimation on a resource constrained 
computing device.

# IMPORTANT NOTES

- Make sure to have v4l-utils installed

## Speeding up LED detection

The original code used a Gaussian blur which is an "expensive" operation as many
pixels have to be read in order to compute the value for the target pixel. 
The Gaussian blur has been replaced with a much cheaper Box blur operation.
However, even the Box blur operation might not be necessary and one can think
about removing it.

Some checks for circularity and width-height distortion have been removed as 
they were deemed not necessary (for indoor use!). Adding them back in should be 
trivial.

## Camera driver inside the MPE node

The original code used ROS to communicate with the camera. The communication was 
optimized by using nodelets.
However, there were two conversion steps from the camera to the ROS message and 
to OpenCV (through CV bridge) which involve two copies from the original camera
buffer. Additionally, the color conversion from RGB to Grayscale was done in 
OpenCV. The color conversion from RGB to grayscale is not a cheap operation as
it involves all pixels.

In order to get full control from the camera image to the MPE node the usb_cam 
implementation was used and modified. usb_cam uses video4linux2 as a backend, which
is a standard linux driver. The usb_cam package was modified and most options 
were removed such that it always chooses YUYV as recording format. The grayscale
image is directly extracted from the Y-data involving no additional steps. After
extracting the grayscale the buffer is never copied again -- the pointer to the 
data is directly passed into a CV::Mat. This also means that the image get's
destroyed after the LED detection is applied, which sometimes shows up as artifacts
when viewing the debug image in `rqt_image_view`.

Another change from the standard usb_cam package is that the parameters have been
added to dynamic_reconfigure which makes it very easy to tune exposure settings 
for example.

## Further improvement directions

If a NVIDIA GPU becomes available to rapyuta, using it could lead to better per-
formance in the Matrix-computation heavy operations (like box-blur).
Also, there is a OpenCV Tegra package which can be installed, and should lead
to a speed up in some OpenCV operations (has not been explored yet).

http://docs.nvidia.com/gameworks/content/technologies/mobile/opencv_tegra_examples.htm

monocular_pose_estimator/src/monocular_pose_estimator_usb_cam.cpp
// This file is part of RPG-MPE - the RPG Monocular Pose Estimator
//
// RPG-MPE is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// RPG-MPE is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with RPG-MPE.  If not, see <http://www.gnu.org/licenses/>.

/*
 * monocular_pose_estimator.cpp
 *
 * Created on: Jul 29, 2013
 * Author: Karl Schwabe
 *         Wolf Vollprecht <wolf.vollprecht@rapyuta-robotics.com>
 */

/** \file monocular_pose_estimator_node.cpp
 * \brief File containing the main function of the package
 *
 * This file is responsible for the flow of the program.
 *
 */

#include "monocular_pose_estimator/monocular_pose_estimator_usb_cam.h"

namespace monocular_pose_estimator {

/**
 * Constructor of the Monocular Pose Estimation Node class
 *
 */
MPENode::MPENode(const ros::NodeHandle& nh, const ros::NodeHandle& nh_private)
    : nh_(nh), nh_private_(nh_private), have_camera_info_(false) {
  // Set up a dynamic reconfigure server.
  // This should be done before reading parameter server values.
  
  dynamic_reconfigure::Server<
      monocular_pose_estimator::MonocularPoseEstimatorConfig>::CallbackType cb_;
  cb_ = boost::bind(&MPENode::dynamicParametersCallback, this, _1, _2);
  dr_server_.setCallback(cb_);

  // Initialize subscribers
  std::string video_device_id("/dev/video0");
  nh_private_.getParam("video_device_id", video_device_id);

  nh_private_.param("camera_brightness", brightness_, -1); // 0-255, -1 "leave alone"
  nh_private_.param("camera_contrast", contrast_, -1); // 0-255, -1 "leave alone"
  nh_private_.param("camera_saturation", saturation_, -1); // 0-255, -1 "leave alone"
  nh_private_.param("camera_sharpness", sharpness_, -1); // 0-255, -1 "leave alone"
  
  // enable/disable autofocus
  nh_private_.param("camera_autofocus", autofocus_, false);
  nh_private_.param("camera_focus", focus_, -1); // 0-255, -1 "leave alone"
  
  // enable/disable autoexposure
  nh_private_.param("camera_autoexposure", autoexposure_, false);
  nh_private_.param("camera_exposure", exposure_, 100);
  nh_private_.param("camera_gain", gain_, -1); // 0-100?, -1 "leave alone"

  video_capture_.start(video_device_id.c_str(), 640, 480, 30);

  setCameraParameters();

  video_capture_.start_capturing();

  // Initialize pose publisher
  pose_pub_ = nh_.advertise<geometry_msgs::PoseWithCovarianceStamped>(
      "estimated_pose", 1);

  // Initialize image publisher for visualization
  image_transport::ImageTransport image_transport(nh_);
  image_pub_ = image_transport.advertise("image_with_detections", 1);

  // Create the marker positions from the test points
  List4DPoints positions_of_markers_on_object;

  // Read in the marker positions from the YAML parameter file
  XmlRpc::XmlRpcValue points_list;
  if (!nh_private_.getParam("marker_positions", points_list)) {
    ROS_ERROR(
        "%s: No reference file containing the marker positions, or the file is "
        "improperly formatted. Use the 'marker_positions_file' parameter in "
        "the launch file.",
        ros::this_node::getName().c_str());
    ros::shutdown();
  } else {
    positions_of_markers_on_object.resize(points_list.size());
    for (int i = 0; i < points_list.size(); i++) {
      Eigen::Matrix<double, 4, 1> temp_point;
      temp_point(0) = points_list[i]["x"];
      temp_point(1) = points_list[i]["y"];
      temp_point(2) = points_list[i]["z"];
      temp_point(3) = 1;
      positions_of_markers_on_object(i) = temp_point;
    }
  }

  trackable_object_.setMarkerPositions(positions_of_markers_on_object);
  ROS_INFO("The number of markers on the object are: %d",
           (int)positions_of_markers_on_object.size());

  std::string camera_info_url;
  if (!nh_private_.getParam("camera_info_url", camera_info_url)) {
    ROS_ERROR("No camera info url set!");
    ros::shutdown();
  } else {
    std::string camera_name;
    camera_calibration_parsers::readCalibration(camera_info_url, camera_name,
                                                cam_info_);
    storeCamInfo();
    ROS_INFO("Camera info loaded for %s", camera_name.c_str());
  }

  int param_setter_tick = 250;
  ros::Rate rate(30);
  while (ros::ok()) {
    run();
    ros::spinOnce();
    rate.sleep();
    if(param_setter_tick == 300) {
      this->setCameraParameters();
      param_setter_tick = 0;
    }
    param_setter_tick++;
  }
}

/**
 * Destructor of the Monocular Pose Estimation Node class
 *
 */
MPENode::~MPENode() {}

/**
 * The callback function that retrieves the camera calibration information
 *
 * \param msg the ROS message containing the camera calibration information
 *
 */
void MPENode::storeCamInfo() {
  // Calibrated camera
  trackable_object_.camera_matrix_K_ = cv::Mat(3, 3, CV_64F);
  trackable_object_.camera_matrix_K_.at<double>(0, 0) = cam_info_.K[0];
  trackable_object_.camera_matrix_K_.at<double>(0, 1) = cam_info_.K[1];
  trackable_object_.camera_matrix_K_.at<double>(0, 2) = cam_info_.K[2];
  trackable_object_.camera_matrix_K_.at<double>(1, 0) = cam_info_.K[3];
  trackable_object_.camera_matrix_K_.at<double>(1, 1) = cam_info_.K[4];
  trackable_object_.camera_matrix_K_.at<double>(1, 2) = cam_info_.K[5];
  trackable_object_.camera_matrix_K_.at<double>(2, 0) = cam_info_.K[6];
  trackable_object_.camera_matrix_K_.at<double>(2, 1) = cam_info_.K[7];
  trackable_object_.camera_matrix_K_.at<double>(2, 2) = cam_info_.K[8];
  trackable_object_.camera_distortion_coeffs_ = cam_info_.D;

  have_camera_info_ = true;
  ROS_INFO("Camera calibration information obtained.");
}

/**
 * The callback function that is executed every time an image is received. It
 *runs the main logic of the program.
 *
 * \param image_msg the ROS message containing the image to be processed
 */
void MPENode::run() {
  // Check whether already received the camera calibration data
  if (!have_camera_info_) {
    ROS_WARN("No camera info yet...");
    return;
  }

  // Get time at which the image was taken. This time is used to stamp the
  // estimated pose and also calculate the position of where to search for the
  // makers in the image
  ros::Time time_of_frame = ros::Time::now();

  // Mat(int _rows, int _cols, int _type, void* _data, size_t _step=AUTO_STEP);
  cv::Mat image =
      cv::Mat(480, 640, CV_8U, (void*) video_capture_.grab_image_pointer());

  const bool found_body_pose =
      trackable_object_.estimateBodyPose(image, time_of_frame.toSec());

  if (found_body_pose)  // Only output the pose, if the pose was updated (i.e. a
                        // valid pose was found).
  {
    // Eigen::Matrix4d transform = trackable_object.getPredictedPose();
    Matrix6d cov = trackable_object_.getPoseCovariance();
    Eigen::Matrix4d transform = trackable_object_.getPredictedPose();

    ROS_DEBUG_STREAM("The transform: \n" << transform);
    ROS_DEBUG_STREAM("The covariance: \n" << cov);

    // Convert transform to PoseWithCovarianceStamped message
    predicted_pose_.header.stamp = time_of_frame;
    predicted_pose_.header.frame_id = "mpe_frame";
    predicted_pose_.pose.pose.position.y = transform(0, 3);
    predicted_pose_.pose.pose.position.x = -transform(1, 3);
    predicted_pose_.pose.pose.position.z = -transform(2, 3);
    auto ned_quat = Eigen::Quaterniond(0.0, -1.0 / std::sqrt(2), 1.0 / std::sqrt(2), 0.0);
    Eigen::Quaterniond orientation = ned_quat *
        Eigen::Quaterniond(transform.block<3, 3>(0, 0));
    // auto orientation = Eigen::Quaterniond(transform.block<3, 3>(0, 0));
    predicted_pose_.pose.pose.orientation.x = orientation.x();
    predicted_pose_.pose.pose.orientation.y = orientation.y();
    predicted_pose_.pose.pose.orientation.z = orientation.z();
    predicted_pose_.pose.pose.orientation.w = orientation.w();

    // Add covariance to PoseWithCovarianceStamped message
    for (unsigned i = 0; i < 6; ++i) {
      for (unsigned j = 0; j < 6; ++j) {
        predicted_pose_.pose.covariance.elems[j + 6 * i] = cov(i, j);
      }
    }

    // Publish the pose
    pose_pub_.publish(predicted_pose_);
  } else {  // If pose was not updated
    ROS_WARN("Unable to resolve a pose.");
  }

  // publish visualization image
  if (image_pub_.getNumSubscribers() > 0) {
    cv::Mat visualized_image = image.clone();
    cv::cvtColor(visualized_image, visualized_image, CV_GRAY2RGB);
    trackable_object_.augmentImage(visualized_image, found_body_pose);

    // Publish image for visualization
    cv_bridge::CvImage visualized_image_msg;
    visualized_image_msg.encoding = sensor_msgs::image_encodings::BGR8;
    visualized_image_msg.image = visualized_image;

    image_pub_.publish(visualized_image_msg.toImageMsg());
  }
}

void MPENode::setCameraParameters() {
  // set camera parameters
  if (brightness_ >= 0)
    video_capture_.set_v4l_parameter("brightness", brightness_);
  if (contrast_ >= 0)
    video_capture_.set_v4l_parameter("contrast", contrast_);
  if (saturation_ >= 0)
    video_capture_.set_v4l_parameter("saturation", saturation_);
  if (sharpness_ >= 0)
    video_capture_.set_v4l_parameter("sharpness", sharpness_);
  if (gain_ >= 0)
    video_capture_.set_v4l_parameter("gain", gain_);

  // check auto white balance
  // if (auto_white_balance_) {
  //   video_capture_.set_v4l_parameter("white_balance_temperature_auto", 1);
  // } else {
  //   video_capture_.set_v4l_parameter("white_balance_temperature_auto", 0);
  //   video_capture_.set_v4l_parameter("white_balance_temperature", white_balance_);
  // }

  // check auto exposure
  if (!autoexposure_) {
    // turn down exposure control (from max of 3)
    video_capture_.set_v4l_parameter("exposure_auto", 1);
    video_capture_.set_v4l_parameter("exposure_absolute", exposure_);
  }

  // check auto focus
  if (autofocus_) {
    video_capture_.set_auto_focus(1);
    video_capture_.set_v4l_parameter("focus_auto", 1);
  } else {
    video_capture_.set_v4l_parameter("focus_auto", 0);
    if (focus_ >= 0)
      video_capture_.set_v4l_parameter("focus_absolute", focus_);
  }
}

/**
 * The dynamic reconfigure callback function. This function updates the variable
 * within the program whenever they are changed using dynamic reconfigure.
 */
void MPENode::dynamicParametersCallback(
    monocular_pose_estimator::MonocularPoseEstimatorConfig& config,
    uint32_t level) {
  ROS_WARN("CHANGING PARAMETERS!");
  trackable_object_.detection_threshold_value_ = config.threshold_value;
  trackable_object_.blur_size_ = config.blur_size;
  trackable_object_.min_blob_area_ = config.min_blob_area;
  trackable_object_.max_blob_area_ = config.max_blob_area;
  trackable_object_.max_width_height_distortion_ =
      config.max_width_height_distortion;
  trackable_object_.max_circular_distortion_ = config.max_circular_distortion;
  trackable_object_.roi_border_thickness_ = config.roi_border_thickness;

  trackable_object_.setBackProjectionPixelTolerance(
      config.back_projection_pixel_tolerance);
  trackable_object_.setNearestNeighbourPixelTolerance(
      config.nearest_neighbour_pixel_tolerance);
  trackable_object_.setCertaintyThreshold(config.certainty_threshold);
  trackable_object_.setValidCorrespondenceThreshold(
      config.valid_correspondence_threshold);

  brightness_ = config.camera_brightness;
  contrast_ = config.camera_contrast;
  saturation_ = config.camera_saturation;
  sharpness_ = config.camera_sharpness;
  focus_ = config.camera_focus;
  exposure_ = config.camera_exposure;
  gain_ = config.camera_gain;

  autofocus_ = config.camera_autofocus;
  autoexposure_ = config.camera_autoexposure;

  setCameraParameters();

  ROS_INFO("Parameters changed");
}

}  // namespace monocular_pose_estimator
